source/rules/tokens.rs

```diff
@@ -0,0 +1,222 @@
+// Tagua VM
+//
+//
+// New BSD License
+//
+// Copyright © 2016-2016, Ivan Enderlin.
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//     * Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//     * Neither the name of the Hoa nor the names of its contributors may be
+//       used to endorse or promote products derived from this software without
+//       specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS AND CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+//! Group of token rules.
+//!
+//! The list of all tokens is provided by the PHP Language Specification in
+//! the [Grammar chapter, Tokens
+//! section](https://github.com/php/php-langspec/blob/master/spec/19-grammar.md#tokens).
+
+use super::super::tokens;
+use super::super::ast::Name;
+
+named!(
+    pub variable<Name>,
+    map_res!(
+        preceded!(
+            tag!(tokens::VARIABLE),
+            name
+        ),
+        variable_mapper
+    )
+);
+
+fn variable_mapper(string: &[u8]) -> Result<Name, ()> {
+    Ok(Name::Variable(string))
+}
+
+named!(
+    pub qualified_name<Name>,
+    chain!(
+        head: alt!(
+            tag!(tokens::NAMESPACE_SEPARATOR)
+          | terminated!(
+                tag!(tokens::NAMESPACE),
+                tag!(tokens::NAMESPACE_SEPARATOR)
+            )
+        )? ~
+        mut accumulator: map_res!(
+            and_not!(name, tag!(tokens::NAMESPACE)),
```

Comment:
Actually, the grammar does not disallow/forbid it. This is even ambiguous, since a *token* is a disjunction between several things and a *name* and a *keyword*, but *keyword* is a subset of *name*.

We can consider that `Foo\class` is invalid because of a limitation due to the nature of the chosen parser. If there is indeed a lexical analysis preceding a syntactic analysis, then yes, `Foo\class` will be invalid. But with a parser combinator, this is possible to have `class` as a classname, like `and` as a method etc. In PHP, this is also very surprising to have `$this->and(…)` as a valid match while `function and` is invalid (calling vs. declaring); keywords are not applied everywhere.

For what I can undertand, a keyword is a special token in specific location in the grammar but it is not “reserved”.

I know that `class` as a classname is invalid in the Zend Engine implementation. I don't know if we should follow this or not. Is it a technical issue in the ZE implementation or is it an editor issue in the specification?

Note that for us, it would be easily fixed with the `exclude` parser macro, with something like:
```rust
exclude!(name, keywords)
```
(we need to declare `keywords`, but tokens are already written).